File: platform/polyglot/confluence_api/py/src/confluence_api/pagination.py
"""
Pagination utilities for the Confluence Data Center REST API.

Supports two pagination strategies:
- Offset-based: standard /content, /space endpoints (start/limit/size)
- Cursor-based: /content/scan endpoint (cursor token in _links.next)
"""

from __future__ import annotations

from collections.abc import AsyncIterator
from typing import Any

from confluence_api.logger import create_logger

log = create_logger("confluence-api", __file__)


async def paginate_offset(
    client: Any,
    endpoint: str,
    params: dict[str, Any] | None = None,
    start: int = 0,
    limit: int = 25,
) -> AsyncIterator[dict[str, Any]]:
    """
    Async generator that yields individual items from offset-based paginated endpoints.

    Confluence offset pagination uses:
        - start: 0-based offset
        - limit: page size
        - size: number of results in current page
        - results: array of items

    The generator automatically fetches subsequent pages until all results are consumed.

    Args:
        client: ConfluenceClient instance (must have .get() method).
        endpoint: API endpoint path (e.g. 'content', 'space').
        params: Additional query parameters to include in each request.
        start: Initial offset (default 0).
        limit: Page size (default 25, Confluence max is typically 250).

    Yields:
        Individual result dicts from each page.
    """
    current_start = start
    base_params = dict(params) if params else {}

    while True:
        page_params = {**base_params, "start": current_start, "limit": limit}
        log.debug("paginate_offset fetching page", {"endpoint": endpoint, "start": current_start, "limit": limit})

        response = client.get(endpoint, params=page_params)

        results: list[dict[str, Any]] = response.get("results", [])
        if not results:
            break

        for item in results:
            yield item

        # Check if there are more pages
        page_size = response.get("size", len(results))
        total_size = response.get("totalSize")

        # If the page returned fewer items than the limit, we are done
        if page_size < limit:
            break

        # If we know total size and have fetched everything, stop
        if total_size is not None and (current_start + page_size) >= total_size:
            break

        # Also check _links.next — Confluence includes it when more pages exist
        links = response.get("_links", {})
        if not links.get("next"):
            break

        current_start += page_size


async def paginate_cursor(
    client: Any,
    endpoint: str,
    params: dict[str, Any] | None = None,
    limit: int = 25,
) -> AsyncIterator[dict[str, Any]]:
    """
    Async generator that yields individual items from cursor-based paginated endpoints.

    Used by /content/scan which returns a cursor token in _links.next rather than
    offset-based navigation. This is more efficient for large result sets.

    Args:
        client: ConfluenceClient instance (must have .get() method).
        endpoint: API endpoint path (e.g. 'content/scan').
        params: Additional query parameters to include in each request.
        limit: Page size (default 25).

    Yields:
        Individual result dicts from each page.
    """
    base_params = dict(params) if params else {}
    cursor: str | None = None

    while True:
        page_params = {**base_params, "limit": limit}
        if cursor is not None:
            page_params["cursor"] = cursor

        log.debug("paginate_cursor fetching page", {"endpoint": endpoint, "cursor": cursor, "limit": limit})

        response = client.get(endpoint, params=page_params)

        results: list[dict[str, Any]] = response.get("results", [])
        if not results:
            break

        for item in results:
            yield item

        # Extract cursor from _links.next
        links = response.get("_links", {})
        next_link: str | None = links.get("next")

        if not next_link:
            break

        # Parse cursor from the next URL query string
        cursor = _extract_cursor(next_link)
        if cursor is None:
            break


def _extract_cursor(next_link: str) -> str | None:
    """
    Extract the cursor parameter from a Confluence _links.next URL.

    The next link is typically a relative URL like:
        /rest/api/content/scan?cursor=...&limit=25

    Args:
        next_link: The next URL from _links.next.

    Returns:
        The cursor value, or None if not found.
    """
    from urllib.parse import parse_qs, urlparse

    try:
        parsed = urlparse(next_link)
        qs = parse_qs(parsed.query)
        cursor_values = qs.get("cursor", [])
        return cursor_values[0] if cursor_values else None
    except Exception:
        log.warning("failed to extract cursor from next link", {"next_link": next_link})
        return None


def build_expand(fields: list[str]) -> str:
    """
    Build an expand parameter value by joining field names with commas.

    Confluence uses the ?expand= query parameter to request additional data
    in responses. Multiple fields are comma-separated.

    Args:
        fields: List of field names to expand (e.g. ['body.storage', 'version', 'ancestors']).

    Returns:
        Comma-separated string suitable for the expand parameter.

    Example:
        >>> build_expand(['body.storage', 'version', 'ancestors'])
        'body.storage,version,ancestors'
    """
    return ",".join(fields)

File: platform/polyglot/confluence_api/mjs/src/pagination.mjs
/**
 * @module pagination
 * @description Pagination utilities for the Confluence Data Center REST API.
 *
 * Confluence uses two pagination strategies:
 *
 * 1. **Offset-based pagination** — Most endpoints use `start` (0-based offset) and
 *    `limit` (page size) parameters. The response includes `results[]`, `start`,
 *    `limit`, and `size` fields. Iteration stops when `size < limit`.
 *
 * 2. **Cursor-based pagination** — The `/content/scan` endpoint (and a few others)
 *    use an opaque `cursor` token. The response includes `_links.next` with the
 *    cursor for the subsequent page. Iteration stops when `_links.next` is absent.
 *
 * Both strategies are exposed as async generators that yield individual items,
 * abstracting away the underlying pagination mechanics.
 *
 * @example
 * // Offset-based: iterate all pages in a space
 * for await (const page of paginateOffset(client, '/rest/api/content', {
 *   spaceKey: 'DEV', type: 'page'
 * })) {
 *   console.log(page.title);
 * }
 *
 * // Cursor-based: scan all content
 * for await (const item of paginateCursor(client, '/rest/api/content/scan')) {
 *   console.log(item.id);
 * }
 */

/**
 * Async generator for offset-based pagination.
 *
 * Fetches pages of results from a Confluence REST endpoint using `start` and `limit`
 * query parameters. Yields individual items from each page's `results` array.
 * Automatically advances the offset until the server returns fewer results than
 * the requested limit, indicating the final page.
 *
 * @template T
 * @param {import('./client/ConfluenceFetchClient.mjs').ConfluenceFetchClient} client
 *   The Confluence fetch client instance.
 * @param {string} endpoint - API endpoint path (e.g. '/rest/api/content').
 * @param {Record<string, unknown>} [params={}] - Additional query parameters
 *   (e.g. spaceKey, type, expand).
 * @param {Object} [options={}] - Pagination options.
 * @param {number} [options.start=0] - Initial offset (0-based).
 * @param {number} [options.limit=25] - Page size (max items per request).
 * @yields {T} Individual result items from each page.
 *
 * @example
 * const client = new ConfluenceFetchClient({ baseUrl, username, apiToken });
 * for await (const page of paginateOffset(client, '/rest/api/content', {
 *   spaceKey: 'DEV',
 *   type: 'page',
 *   expand: 'version,space',
 * }, { limit: 50 })) {
 *   console.log(page.id, page.title);
 * }
 */
export async function* paginateOffset(client, endpoint, params = {}, { start = 0, limit = 25 } = {}) {
  let currentStart = start;

  while (true) {
    const queryParams = {
      ...params,
      start: currentStart,
      limit,
    };

    /** @type {{ results?: T[], start?: number, limit?: number, size?: number }} */
    const response = await client.get(endpoint, { queryParams });

    const results = response.results ?? [];
    for (const item of results) {
      yield item;
    }

    // If we received fewer results than the limit, we've reached the last page
    const size = response.size ?? results.length;
    if (size < limit) {
      break;
    }

    currentStart += size;
  }
}

/**
 * Async generator for cursor-based pagination.
 *
 * Fetches pages of results from a Confluence REST endpoint (typically
 * `/content/scan`) using an opaque cursor token. Yields individual items from
 * each page's `results` array. Automatically follows `_links.next` until no
 * further cursor is provided.
 *
 * @template T
 * @param {import('./client/ConfluenceFetchClient.mjs').ConfluenceFetchClient} client
 *   The Confluence fetch client instance.
 * @param {string} endpoint - API endpoint path (e.g. '/rest/api/content/scan').
 * @param {Record<string, unknown>} [params={}] - Additional query parameters
 *   (e.g. expand, cql).
 * @param {Object} [options={}] - Pagination options.
 * @param {number} [options.limit=25] - Page size (max items per request).
 * @yields {T} Individual result items from each page.
 *
 * @example
 * for await (const content of paginateCursor(client, '/rest/api/content/scan', {
 *   expand: 'version',
 * }, { limit: 100 })) {
 *   console.log(content.id, content.type);
 * }
 */
export async function* paginateCursor(client, endpoint, params = {}, { limit = 25 } = {}) {
  let cursor = undefined;

  while (true) {
    const queryParams = {
      ...params,
      limit,
    };

    if (cursor) {
      queryParams.cursor = cursor;
    }

    /** @type {{ results?: T[], _links?: { next?: string } }} */
    const response = await client.get(endpoint, { queryParams });

    const results = response.results ?? [];
    for (const item of results) {
      yield item;
    }

    // Extract cursor from _links.next
    // Confluence returns _links.next as a relative URL with query params including cursor
    const nextLink = response._links?.next;
    if (!nextLink) {
      break;
    }

    // Parse the cursor from the next link URL
    try {
      // _links.next may be a relative path like "/rest/api/content/scan?cursor=abc123&limit=25"
      const nextUrl = new URL(nextLink, 'http://placeholder');
      const nextCursor = nextUrl.searchParams.get('cursor');
      if (!nextCursor) {
        break;
      }
      cursor = nextCursor;
    } catch {
      // If we can't parse the next link, stop pagination
      break;
    }
  }
}

/**
 * Build a comma-separated `expand` parameter value from an array of field names.
 *
 * Confluence REST APIs use the `expand` query parameter to include additional
 * data in responses. This utility joins an array of field names with commas,
 * handling null/undefined inputs gracefully.
 *
 * @param {string|string[]|null|undefined} fields - Field names to expand.
 * @returns {string|undefined} Comma-separated string, or undefined if no fields provided.
 *
 * @example
 * buildExpand(['body.storage', 'version', 'space']);
 * // => 'body.storage,version,space'
 *
 * buildExpand('body.storage');
 * // => 'body.storage'
 *
 * buildExpand(null);
 * // => undefined
 */
export function buildExpand(fields) {
  if (!fields) return undefined;
  if (typeof fields === 'string') return fields;
  if (Array.isArray(fields) && fields.length > 0) {
    return fields.join(',');
  }
  return undefined;
}
